plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,434 @@
+#include "wayfire/core.hpp"
+
+#include "wayfire/seat.hpp"
+#include "wayfire/scene-input.hpp"
+#include "wayfire/scene-operations.hpp"
+#include "wayfire/unstable/wlr-surface-node.hpp"
+#include "wayfire/unstable/wlr-surface-controller.hpp"
+#include "wayfire/output.hpp"
+#include "wayfire/output-layout.hpp"
+#include "wayfire/signal-definitions.hpp"
+
+#include <wayfire/plugin.hpp>
+#include <wayfire/util.hpp>
+#include <wayfire/nonstd/wlroots-full.hpp>
+#include <wayfire/plugins/common/simple-text-node.hpp>
+#include <wayfire/util/log.hpp>
+
+enum lock_state
+{
+    LOCKING,
+    LOCKED,
+    UNLOCKED,
+    DESTROYED,
+    ZOMBIE,
+} state;
+
+class wf_session_lock_plugin : public wf::plugin_interface_t
+{
+    class wayfire_session_lock;
+
+    class lock_surface_node_t : public wf::scene::wlr_surface_node_t
+    {
+      public:
+        lock_surface_node_t(wayfire_session_lock *lock,
+            wlr_session_lock_surface_v1 *lock_surface,
+            wf::output_t *output) :
+            wf::scene::wlr_surface_node_t(lock_surface->surface, true /* autocommit */),
+            lock(lock),
+            lock_surface(lock_surface),
+            output(output),
+            interaction(std::make_unique<lock_surface_keyboard_interaction>(lock_surface->surface,
+                lock_surface->output))
+        {
+            lock_surface_destroy.set_callback([this] (void *data)
+            {
+                wf::wlr_surface_controller_t::try_free_controller(this->lock_surface->surface);
+                wf::scene::remove_child(shared_from_this());
+                lock_surface_destroy.disconnect();
+                this->lock->surface_destroyed(this->output);
```

Comment:
Re lifetimes: I would personally set the destroy callback in the wayfire_session_lock class (this way, you don't have to pass the `lock` parameter.